### PR TITLE
[WIP] aflplusplus: 2.64c -> 4.05c

### PR DESCRIPTION
**Pull request by lf-**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lf-** [comment]
ping maintainers: @Mindavi @risicle

Hi! I've mostly updated afl to 4.05c, and haven't really tested it too much. The test-suite almost works, but there's some breakage in cmocka related unit tests, as well as tests using afl-cmin.bash (those just need some patching).

I have run out of time to work on this as it took longer than expected, however I am posting it here as a starting point at least.

**Comment by Mindavi** [comment]
Thanks for doing this initial work! It does seem like it's already in a good shape, if I find some time soon I might pick it up and get it over the finish line. I find it kind of hard to maintain this since I haven't used the wine / blackbox fuzzing ever and we try to support it in this package. We should really have a more extensive test suite with the various forms that afl supports, their own suite is (was?) rather limited.

Personally I'd rather get this up to date without some optional things if that means we can at least update it. Wdyt @risicle ?

File: pkgs/tools/security/aflplusplus/default.nix [hunk 1]
```diff
@@ -105,17 +97,25 @@ let
...
+        --replace '../libcompcov.so' '`$out/bin/get-afl-qemu-libcompcov-so`'
+      substituteInPlace test/test-libextensions.sh \
         --replace '../libdislocator.so' '`$out/bin/get-libdislocator-so`' \
         --replace '../libtokencap.so' '`$out/bin/get-libtokencap-so`'
-      perl -pi -e 's|(?<!\.)(?<!-I)(\.\./)([^\s\/]+?)(?<!\.c)(?<!\.s?o)(?=\s)|\$out/bin/\2|g' test/test.sh
-      cd test && ./test.sh
+      substituteInPlace test/test-llvm.sh \
+        --replace '../afl-cmin.bash' '`$out/bin/get-libdislocator-so`' \
+        --replace '../libtokencap.so' '`$out/bin/get-libtokencap-so`'
+      # what the fuck
```
Comment (Mindavi) [inline pkgs/tools/security/aflplusplus/default.nix:114]: Let's not do that :stuck_out_tongue_winking_eye:

**Review by Mindavi — CHANGES_REQUESTED** [review]
Some initial comments, there's quite a bit of work left to get this working since the build system is not too nice for nixpkgs.

**Comment by Mindavi** [comment]
Been hacking at this at https://github.com/Mindavi/nixpkgs/tree/aflplusplus/4.05c and made some nice progress. But still a lot is not working yet.

**Comment by lopsided98** [comment]
I tested @Mindavi's version (updated to 4.06c) and it works correctly for my basic use case (using `afl-clang-lto`).

**Comment by Mindavi** [comment]
Sounds good! Let me open a new PR with my branch, rebased on top of master. Then we can finally get this update in :). I can live with some features not fully working, I'm not using them myself anyway. If anyone ever needs them, hopefully they can improve the support again.

**Comment by Mindavi** [comment]
Follow-up PR: https://github.com/NixOS/nixpkgs/pull/234326

**Pull request closed without merging**